lib/common/output_xml.c

```diff
@@ -0,0 +1,258 @@
+/*
+ * Copyright 2019 the Pacemaker project contributors
+ *
+ * The version control history for this file may have further details.
+ *
+ * This source code is licensed under the GNU Lesser General Public License
+ * version 2.1 or later (LGPLv2.1+) WITHOUT ANY WARRANTY.
+ */
+
+#ifndef _GNU_SOURCE
+#  define _GNU_SOURCE
+#endif
+
+#include <ctype.h>
+#include <stdarg.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <crm/crm.h>
+#include <crm/common/output.h>
+#include <crm/common/xml.h>
+
+typedef struct xml_private_s {
+    xmlNode *root;
+    GQueue *parent_q;
+} xml_private_t;
+
+static char *
+mangle_element_id(const char *in) {
+    char *out = NULL;
+
+    /* XML elements may not start with the string "XML" in any case, so add
+     * a leading underscore if it's seen.
+     */
+    if (strncasecmp(in, "XML", 3) == 0) {
+        out = crm_strdup_printf("_%s", in);
+    } else {
+        out = strdup(in);
+    }
```

Comment:
This seems needlessly restrictive, since in combination with
https://www.w3.org/TR/xml-names/#ns-decl, it seems to be only
a potential clash (nothing like that happens and on the element
name level nad libxml2 is fully OK with such) only with
non-namespaced elements/attributes.

Solution would be to namespace the elements, or reconsidering whether
custom values/names based identifiers should really be modeled using
name-of-element-or-attribute approach, since there are more apparent
downsides of that (client never wants a skewed view).  If it's for
internally provided values only, it would perhaps be more sensible
to trap on incompatible values rather than playing this obscenity?

Reply:
Also if see correctly, there's no reason to apply this function with
the current schema, right?

Reply:
I don't see anything wrong with the current approach. It's certainly not an obscenity; converting arbitrary strings into a more limited lexical space is a useful practical capability that doesn't need to stretch into the limits of idealized perfection. Future-proofing is a good idea, too.

However, looking at the where we use this, it may have become unnecessary since we added the custom message model. We only use it for the non-custom, generic messaging functions common to all formatters, and I think it might make sense to use generic element names for that case, e.g. list and list_item

Reply:
This is also how text strings get turned into element IDs in a way that the same text strings can be used in both the XML and the text formatters.  For instance:
```
out->begin_list(out, "Fencing history", "event", "events");
```
For the text formatter, "Fencing history" will be the header for the list in the text output.  For the XML formatter, the mangle function will convert that into something appropriate.

Reply:
Right, but I think something like &lt;list name="Fencing history"&gt; would perhaps be more useful (e.g. the eventual user could use that directly as a heading). For custom messages it makes sense to use custom element names, but for the generic functions, the content will always have the same structure. (In this case we have a custom message for the list items rather than using the generic list item message, which I think highlights the distinction I'm getting at.)

Reply:
To express that as RNG, I wouldn't bother constraining lists to have matching names and items, I'd have a single list element with all the various list item possibilities as choices. That will be more future-proof and maintainable.

Reply:
I am still unclear about the intentions, and I appreciate you have more
clear picture, but let me get this straight, generating ad-hoc elements
or attribute names is indeed obscene in the structured data storage and
retriaval sense, since it prevents any sort of schema/regularity
(= predictability at the receiver) enforcement (which destroys the idea
the structured output comes from), and in that light, the referred
function is either superfluous or shall be modified as internal sanity
enforcer only, IMHO.

Imagine:

```
<Supported-standards>
  <item>ocf</item>
  ...
</items>
```

or

```
<items Supported-standards="crm_resource --list-standards">
  <item>ocf</item>
  ...
</items>
```

Yes, valid and doable.  Usable?  Not so much.
Changing the alleged table header will wipe the previous effort
on the receiver's side.


On the other hand:

```
<items rel="resources" ref="@standards">
  <caption>Supported standards</caption>
  <item>ocf</item>
  ...
</items>
```

(per https://bugs.clusterlabs.org/show_bug.cgi?id=5376#c2,
renamed `list` to `items`, since there's no ordering!)

Caption changes?  No prob.

Even custome messages shall follow the meta-data continuity like
that, forget about ad-hoc inventions, please.  Rather, let's
settle on the main possibly recursively nested container data types
as I've tried to encourage in the referenced bug comment:

```
<items/>
<items ordered="true"/>
<data/>  <!-- akin to <[[CDATA]]>? -->
<request/>
<response/>
<command/>
```

Reply:
None of it is ad hoc; dynamically generated and/or custom does not mean non-schema-enforced. Everything is specified in the schema except for quoted XML content (like resource agent meta-data output).

I don't have a preference for whether the description is an attribute or a sub-element.

The intent is that the caller will call begin_list() at the beginning of a list of items, and end_list() at the end. For the actual items, it might use the generic list_item() method (if the items are simple strings) or a custom message type (if distinct formatting or greater structure is desired).

"List" doesn't imply ordering, so I don't think we need anything special there. No one's going to write a higher-level program that checks for "ordered" in every output and behaves differently accordingly. They'll either always pass the items through as-is or always order them themselves according to their desired criteria.

I'm not sure what we would gain from rel/ref attributes. Higher-level programs probably wouldn't use anything like that; they're just looking for a more reliable format than user-friendly text.

The only main data type is "pacemaker-result"; everything else is up to the individual tool. The generic message functions are provided only for convenience, for common, non-tool-specific types of messages. It's expected that each tool will define its own set of messages. A good comparison is the gettext model, where each output format here is comparable to a localization in gettext.

Reply:
A key point is that the overarching goal here is not to find the ideal structural description of the superset of all pacemaker output, but a way for pacemaker tools to provide both user-friendly text output to end users and easily parseable output for higher-level tools.

Reply:
> A key point is [...] to find [...] a way for pacemaker tools to
> provide both user-friendly text output to end users and *easily
> parseable output for higher-level tools*.

Exactly, and on-the-fly made up elements do not fit into this picture,
so I argue the `mangle_element_id` function has principally no use,
now or in the future, since every key structural naming shall rather
be static.  Is it clearer now?

Reply:
If someone created an output message using "a*b", then "a-b" would be the fixed, immutable, well-documented, enforced-by-schema element name. Whether in the C code it's a constant string or generated by a function is irrelevant.

Consider a tool that had a particular message with the possibilities "resource-agent" and "fence-agent", and it created that via a function using "%s-agent", agent_type. It wouldn't make the element names any less fixed.

Reply:
Why would somebody do that?

For human output, identifiers are irrelevant, for machine output, they
are crucial.  Putting these two purposes into a single sack would be
unfortunate, despite doable with such fragile bridges from one domain
to the other.  Not everything doable is in line with the goals.

It's like when you'd know that GCC supports identifiers of unlimited
length -- would you pursue that to make the meaning of the carried
values crystal clear?

The only "inconvenience" being fixed with that schema is the necessity
to provide two separate inputs.  Is that really going to matter that
much, or is this rather an unjustified what-if way of thinking?
That's rhetorical question.  I don't know the roadmap here.

Reply:
I think we're going to move away from the need for this function, but to give the reasoning, the caller doesn't need to know what output format has been selected. The caller might pass something like "Fence history", which might get printed directly by the text formatter but converted into a fence-history element name or ID attribute by the XML formatter.